toolchain/semantics/semantics_node_stack.h

```diff
@@ -124,9 +155,9 @@ class SemanticsNodeStack {
   // Also works with name_id in the union due to type compatibility.
   auto RequireSoloParseNode(Entry entry) -> void;
 
-  // Requires an entry to have a valid node_id.
-  // Also works with name_id in the union due to type compatibility.
-  auto RequireNodeId(Entry entry) -> void;
+  // Requires an entry to have a valid id.
+  // Also works with all items in the union due to type compatibility.
```

Comment:
If you want to type-pun between union members, please include a `static_assert` that those three `...Id` types are standard-layout types. (Ideally we should be checking `is_layout_compatible` between them, but that's not available until C++20.)

Please also add a comment in the definition of `RequireValidId` that you're intentionally reading from an inactive union member, or move this comment there -- the fact that this does type punning internally is an implementation detail, not part of the interface.